kyo-reactive-streams/shared/src/main/scala/kyo/interop/reactive-streams/StreamSubscriber.scala

```diff
@@ -0,0 +1,209 @@
+package kyo.interop.reactivestreams
+
+import StreamSubscriber.*
+import kyo.*
+import org.reactivestreams.*
+
+final class StreamSubscriber[V] private (
+    bufferSize: Int
+)(
+    using
+    allowance: AllowUnsafe,
+    frame: Frame
+) extends Subscriber[V]:
+
+    private enum UpstreamState derives CanEqual:
+        case Uninitialized                                                               extends UpstreamState
+        case WaitForRequest(subscription: Subscription, items: Chunk[V], remaining: Int) extends UpstreamState
+        case Finished(reason: Maybe[Throwable], leftOver: Chunk[V])                      extends UpstreamState
+    end UpstreamState
+
+    private val state = AtomicRef.Unsafe.init(
+        UpstreamState.Uninitialized -> Maybe.empty[Fiber.Promise.Unsafe[Nothing, Unit]]
+    )
+
+    private def throwIfNull[A](b: A): Unit = if isNull(b) then throw new NullPointerException()
+
+    override def onSubscribe(subscription: Subscription): Unit =
+        throwIfNull(subscription)
+        var sideEffect: () => Unit = () => ()
```

Comment:
would it be possible to avoid the mutable value by using `state.get` and `state.set` directly instead of `state.update`?

Reply:
yes its can be done, but It'll need to be converted into loop method for correctness

Reply:
Makes sense. If a `Loop` is too cumbersome, another option is introducing an `update` that allows retuning a "tail" computation